### ISSUE TITLE
[FEATURE_REQUEST] /awaittrigger

### ISSUE THREAD

**Issue by Technologicat** [SillyTavern/SillyTavern#1777]
### Have you searched for similar requests?

Yes

### Is your feature request related to a problem? If so, please describe.

The `/trigger` slash command is asynchronous, which is fine at the end of a Quick Reply, but if you would like to do several `/send`/`/trigger` pairs in the same QR where later steps depend on the AI's response to previous ones, you currently have to do the following to emulate one synchronous `/send`/`/trigger` pair:

```
/setvar key=tmpPrompt The script's canned reply goes here. The script acts on the behalf of the user. |
/gen lock=on {{getVar::tmpPrompt}} |
/setvar key=tmpOutput |
/send {{getvar::tmpPrompt} |
/sendas name="{{char}}" {{getvar::tmpOutput}} |
```

That is, first generate synchronously, and once it completes, then push both the prompt and the reply into the chat log manually. And then you'll have to remember to `/flushvar` the `tmpPrompt` and `tmpOutput` later...

Also, a drawback of this workaround is that it acts as if streaming mode was off.

### Describe the solution you'd like

A slash command for QRs that would have the effect of "await trigger", in JavaScript parlance, so that the script waits while the AI is generating, and resumes once the AI's reply is complete.

This should respect the streaming mode setting, to make it possible to see the AI's reply while it is being generated.

Also, it should probably have a `lock` option, like `/gen` does, to avoid messing up the script's processing by accidental UI inputs.

This would shorten the above example to:

```
/send The script's canned reply goes here. The script acts on the behalf of the user. |
/awaittrigger lock=on
```

also with no need to explicitly `/flushvar` anything later.

### Describe alternatives you've considered

As a zeroth-order approximation, an `{{awaittrigger}}` macro that expands into the workaround code above.

### Additional context

Here's an actual example of the kind of scripted task where this is useful, implemented using the workaround above. The following QR condenses the single most important point from a scientific abstract into one sentence, in three steps.

For this particular example, one might be able to achieve a similar effect with a general summarization script, but the point is that this example breaks down the task into multiple steps - which is a strategy that helps LLMs perform complex tasks in general.

To use this, assign the script to a Quick Reply button. Then copy and paste an abstract [such as from here](https://arxiv.org/abs/2311.17409) into the message field, and instead of sending the message normally, run this QR.

```
/if left={{input}} right="" rule=eq "/echo severity=warning Empty input. Please paste an abstract first. \| /abort" |
/setvar key=userInput {{input}} |
/echo severity=info Summarizing abstract, please wait [step 1 of 3] |
/setvar key=summarizationPrompt Here is the abstract from a scientific paper. How would you summarize the main points, according to the abstract? The summary should be shorter than the original abstract, at most 100 words. If you need to use formatting in your answer, use Markdown.{{newline}}{{newline}}**Abstract**{{newline}}{{newline}}{{getvar::userInput}}{{newline}} |
/flushvar userInput |
/gen lock=on {{getvar::summarizationPrompt}} |
/setvar key=summarizationOutput |
/send {{getvar::summarizationPrompt}} |
/sendas name="{{char}}" {{getvar::summarizationOutput}} |
/echo severity=info Summarizing abstract, please wait [step 2 of 3] |
/setvar key=summarizationPrompt Great! Can you make that summary shorter? |
/gen lock=on {{getvar::summarizationPrompt}} |
/setvar key=summarizationOutput |
/send {{getvar::summarizationPrompt}} |
/sendas name="{{char}}" {{getvar::summarizationOutput}} |
/echo severity=info Summarizing abstract, please wait [step 3 of 3] |
/setvar key=summarizationPrompt Great, thanks. Can you further summarize just the most important main point, into one sentence? |
/gen lock=on {{getvar::summarizationPrompt}} |
/setvar key=summarizationOutput |
/send {{getvar::summarizationPrompt}} |
/sendas name="{{char}}" {{getvar::summarizationOutput}} |
/flushvar summarizationPrompt |
/flushvar summarizationOutput |
/echo severity=info Done!
```

In this use case, it's crucial for verifiability purposes that the intermediate results are saved into the chat history, as if this was a regular conversation with the AI. That way, if one of my auto-generated one-sentence summaries turns out to be inaccurate, it's possible to look at the chat logs to debug what went wrong.

And yes, I am indeed treating the LLM as an exotic virtual machine for text processing, programmable in English. That's an interesting category of use cases that has practical applications. :)

### Priority

Low (Nice-to-have)

### Is this something you would be keen to implement?

None

**Comment by Technologicat** [comment]
Just as a clarification - this is very low priority, as the workaround works fine, and this is probably a rather rare use case (but please correct me if I'm wrong!).

Also, there was a thinko in my alternative suggestion. Obviously, considering how macros work, if this was to be implemented as a macro, expanding to the workaround code as-is won't work, because the prompt needs to go to the `/send` on the previous line. *Mutatis mutandis*.

In case anyone is wondering about the Timelines references, to save a click or two: the point was that any scripts using canned replies, such as the workaround example posted here, used to cause issues in graph connectivity and layout in Timelines due to converging timelines if the canned reply was sent at the same chat depth in several chats. With those PRs merged, now Timelines works fine with scripts like this.

**Comment by Cohee1207** [comment]
Try this out:
```/send Hello | /trigger await=1 | /echo Done```

**Comment by Technologicat** [comment]
Works for me, thanks!

For those playing at home, here's the same script as above, but using `/trigger await=1`:

```
/if left={{input}} right="" rule=eq "/echo severity=warning Empty input. Please paste an abstract first. \| /abort" |
/setvar key=userInput {{input}} |
/echo severity=info Summarizing abstract, please wait [step 1 of 3] |
/send Here is the abstract from a scientific paper. How would you summarize the main points, according to the abstract? The summary should be shorter than the original abstract, at most 100 words. If you need to use formatting in your answer, use Markdown.{{newline}}{{newline}}**Abstract**{{newline}}{{newline}}{{getvar::userInput}}{{newline}} |
/flushvar userInput |
/trigger await=1 |
/echo severity=info Summarizing abstract, please wait [step 2 of 3] |
/send Great! Can you make that summary shorter? |
/trigger await=1 |
/echo severity=info Summarizing abstract, please wait [step 3 of 3] |
/send Great, thanks. Can you further summarize just the most important main point, into one sentence? |
/trigger await=1 |
/echo severity=info Done!
```
